### ISSUE TITLE
TTNN Generic MLIR Op

### ISSUE THREAD

**Issue by nsmithtt** [tenstorrent/tt-mlir#201]
## TTIR Generic

We want to start building out the compiler side of lowering to a generic op.  As a primer please refer to the linalg generic op, ttir follows the same structure: https://mlir.llvm.org/docs/Dialects/Linalg/#linalggeneric-linalggenericop

`./build/bin/ttmlir-opt --ttir-generic test/ttmlir/Dialect/TTIR/test_generic.mlir`:
```
#any_device = #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>
#map = affine_map<(d0, d1) -> (d0, d1)>
#parallel = #tt.iterator_type<parallel>
module attributes {torch.debug_module_name = "_lambda", tt.system_desc = #tt.system_desc<[{arch = <wormhole_b0>, grid = 8x8, l1_size = 1048576, num_dram_channels = 12, dram_channel_size = 1048576}], [0], [<pcie|host_mmio>], [<0, 0, 0, 0>]>} {
  func.func @forward(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
    %0 = tensor.empty() : tensor<64x128xf32>
    %1 = "ttir.generic"(%arg0, %arg1, %0) <{grid = #tt.grid<1x1>, indexing_maps = [#map, #map, #map], iterator_types = [#parallel, #parallel], operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any_device, #any_device, #any_device]}> ({
    ^bb0(%arg2: tensor<64x128xf32>, %arg3: tensor<64x128xf32>, %arg4: tensor<64x128xf32>):
      %2 = "ttir.kernel"(%arg2, %arg3, %arg4) <{kind = @eltwise, op = @mulitply, operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
      "ttir.yield"(%2) : (tensor<64x128xf32>) -> ()
    }) : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
    return %1 : tensor<64x128xf32>
  }
}
```

This op is composed of 2 primary structural components:
1. The top level `ttir.generic` this has attributes:
    - `indexing_maps`: a list of AffineMapAttr, one AffineMapAttr per each input and output view.
    - `iterator_types`:  an ArrayAttr specifying the type of the enclosing loops. Each element of the list represents and iterator of one of the following types: parallel, reduction, window.
    - These 2 attributes together describe how the operation over the tensor can legally be parallelized / divided.  For this issue, let's only consider `iterator_type`, `"parallel"`, which mean we trivially visit each tensor element in isolation, i.e. element wise.  From these attributes alone I think we can completely infer the data movement for the op, in the case of element wise there is none, but it'll also tell the optimizer how the tensors are legally allowed to be parallelized on a grid.
2. The inner region, we have a lambda-like body with a basic block named `^bb0`.  I like to think of this as the high level implementation of a kernel, and is effectively what each core is executing in parallel.
    - `"ttir.kernel"`: This is an opaque op, can be thought of like an FFI call to some kernel that's hand written somewhere and will be linked in.  I think this should be the first step, we'll have kernels written by hand in C++ that conform to an interface such that the compiler can autogenerate the data movement to bring the data to it.
    - `"ttir.yield"`: Nothing super interesting here, mlir requires us to define terminator ops to end a basic block.

## Lowering to TTNN Generic

Ok, so there actually has already been a decent amount of POC done demonstrating the lowering of a `ttir.generic` op, it's just that it was done inside of the TTMetal dialect, but I think many of the same concepts will carry over to the `ttnn` version.

Again let's look at a simple test, we can run a TTMetal dialect test:
```bash
./build/bin/ttmlir-opt --ttir-generic --ttir-layout --ttir-generic-region-operands-to-memref --ttir-allocate --convert-ttir-to-ttmetal test/ttmlir/Dialect/TTMetal/simple_multiply.mlir
```

This result is quite verbose, so I only stripped out the "generic" op that's called `"ttmetal.dispatch"` in the TTMetal dialect.
```
    %5 = "ttmetal.dispatch"(%1, %3, %4) <{core_ranges = [#ttmetal.core_range<0x0, 1x1>, #ttmetal.core_range<0x0, 1x1>, #ttmetal.core_range<0x0, 1x1>], operandSegmentSizes = array<i32: 2, 1>, operand_cb_port_mapping = [0, 1, 2], threadTypes = [#ttkernel.thread<noc0>, #ttkernel.thread<noc1>, #ttkernel.thread<tensix>]}> ({
    ^bb0(%arg2: !ttkernel.cb<0, 0, memref<64x128xf32, #l1_>>, %arg3: !ttkernel.cb<0, 1, memref<64x128xf32, #l1_>>, %arg4: !ttkernel.cb<0, 2, memref<64x128xf32, #l1_>>):
      "ttkernel.cb_push_back"(%arg2) : (!ttkernel.cb<0, 0, memref<64x128xf32, #l1_>>) -> ()
      "ttkernel.return"() : () -> ()
    }, {
    ^bb0(%arg2: !ttkernel.cb<0, 0, memref<64x128xf32, #l1_>>, %arg3: !ttkernel.cb<0, 1, memref<64x128xf32, #l1_>>, %arg4: !ttkernel.cb<0, 2, memref<64x128xf32, #l1_>>):
      "ttkernel.cb_push_back"(%arg3) : (!ttkernel.cb<0, 1, memref<64x128xf32, #l1_>>) -> ()
      "ttkernel.return"() : () -> ()
    }, {
    ^bb0(%arg2: !ttkernel.cb<0, 0, memref<64x128xf32, #l1_>>, %arg3: !ttkernel.cb<0, 1, memref<64x128xf32, #l1_>>, %arg4: !ttkernel.cb<0, 2, memref<64x128xf32, #l1_>>):
      "ttkernel.builtin"(%arg2, %arg3, %arg4) <{kind = @eltwise, op = @mulitply}> : (!ttkernel.cb<0, 0, memref<64x128xf32, #l1_>>, !ttkernel.cb<0, 1, memref<64x128xf32, #l1_>>, !ttkernel.cb<0, 2, memref<64x128xf32, #l1_>>) -> ()
      "ttkernel.return"() : () -> ()
    }) : (tensor<64x128xf32, #layout1>, tensor<64x128xf32, #layout1>, tensor<64x128xf32, #layout1>) -> tensor<64x128xf32, #layout1>
```

So breaking this down again, it has a fairly similar structure to `ttir.generic` except now it's been decomposed into 3 regions, corresponding to 3 threads, noc0, noc1 and tensix.  Notice also these regions have been decomposed into the TTKernel dialect which strives to model as closely as possible the metal [compute](https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/include/compute_kernel_api.h) + [dataflow](https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/hw/inc/dataflow_api.h) programming model.

TTKernel also supports going through the EmitC dialect, so it is expected that they will be converted to C++ and then passed into the TTNN runtime interface to be compiled by the GCC toolchain.

Some things to call out:
- Ops from the TTKernel dialect should be enforced to only be legal inside of regions of `ttmetal.dispatch` and `ttnn.generic`.
- `ttkernel.cb<0, 0, memref<64x128xf32, #l1_>>`: Metal uses circular buffers (CBs) for sharing L1 data between risc threads.  Here CBs replace all the tensor arguments and represent the working shard of the tensor for that core.  The embedded `memref` defines the shape of the tensor shard for each core.  Generally noc threads will be pushing data to inputs tensor's CB, and tensix thread will be popping from it, doing work, and then pushing to the output CB.
- `"ttkernel.builtin"`: Just like `ttir.kernel`, this is an FFI function call to some kernel loop that's been hand written.  For now the compiler could just pattern match from a set of hand written kernel functions based on the `kind` and `op` attributes and then inline the function during C++ generation.

## Tasks

So breaking down the above into tasks:
- #202
- #203
- #204 
- Coordinate with issue #111

**Comment by kmitrovicTT** [comment]
Writing an update on how far we have come with this.

## ttnn progress

As a starting point, @mbezuljTT devised and @pjanevskiTT continued to work on https://github.com/tenstorrent/tt-mlir/issues/111, where they implemented 

```
struct GenericOp {
    static Tensor operator()(
        const std::vector<Tensor> &input_tensors, 
        Tensor& output_tensor, 
        const GenericOpDeviceOperation::operation_attributes_t& operation_attributes);
};
```

where 

```
struct operation_attributes_t {
    std::unordered_map<uint8_t, circular_buffer_attributes_t> circular_buffer_attributes;
    std::vector<data_movement_attributes_t> data_movement_attributes;
    std::vector<compute_attributes_t> compute_attributes;
};
```
and so on, not going into details here. 

Caller of `ttnn::generic_op` has to provide input and output tensors (respecting Destination Passing Style we use in MLIR) as well as a struct which contains various settings (like settings for circular buffers, paths to data movement and compute kernels, etc). The idea is that user/caller will implement their own kernels (maybe even just compute kernels, if data movement can be automatically generated, which is an idea for future work) and some amount of configuration (which is currently in this POC not insignificant, all of it is hidden behind a concrete function call in `ttnn` for example `ttnn::add`) which will be sufficient to run their programs on our HW.

## ttmlir progress

On my [branch](https://github.com/tenstorrent/tt-mlir/tree/kmitrovic/ttnn_generic) I implemented structs contained in `operation_attributes_t` in `TT` dialect, a `ttnn.generic` op and `ttir.external_generic` op (since `ttir.generic` was already in use for Nicks POC for lowering to `ttmetal`), and integrated a pass in `--ttir-to-ttnn-backend-pipeline` which enables us to lower

```
#any_device = #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>
module attributes {} {
  func.func @forward(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
    %0 = tensor.empty() : tensor<64x128xf32>

    %1= "ttir.generic"(%2, %4, %5) <{
      circular_buffer_attributes = [
        #tt.circular_buffer_attributes<c_in0, <0x0, 6x6>, 4096, 2048, bf16>, 
        #tt.circular_buffer_attributes<c_in1, <0x0, 6x6>, 4096, 2048, bf16>, 
        #tt.circular_buffer_attributes<c_out0, <0x0, 6x6>, 4096, 2048, bf16>
      ], 
      compute_attributes = [
        #tt.compute_attributes<
          <0x0, 6x6>, 
          "ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/compute/eltwise_binary_kernel.cpp", 
          <hifi4, false, false, false, <1>, 
          {ELTWISE_OP = "add_tiles", ELTWISE_OP_TYPE = "EltwiseBinaryType::ELWADD"}>>
      ], 
      data_movement_attributes = [
        #tt.data_movement_attributes<
          <0x0, 6x6>, 
          "ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/dataflow/reader_binary_interleaved_start_id.cpp", 
          <reader, <1, 1>>>, 
        #tt.data_movement_attributes<
          <0x0, 6x6>, 
          "ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/dataflow/writer_unary_interleaved_start_id.cpp", 
          <writer, <1, 1>>>
      ], 
      operandSegmentSizes = array<i32: 2, 1>
    }> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32) -> tensor<64x128xf32>

    return %1 : tensor<64x128xf32>
  }
}

```

to

```
#l1_ = #tt.memory_space<l1>
#system = #tt.memory_space<system>
#layout = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #system>>
#layout1 = #tt.layout<(d0, d1) -> (d0, d1), undef, <8x8>, memref<8x16xf32, #system>>
#layout2 = #tt.layout<(d0, d1) -> (d0, d1), undef, <8x8>, memref<8x16xf32, #l1_>>
module attributes {tt.device = #tt.device<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>, tt.system_desc = #tt.system_desc<[{arch = <wormhole_b0>, grid = 8x8, l1_size = 1499136, num_dram_channels = 12, dram_channel_size = 1073741824, noc_l1_address_align_bytes = 16, pcie_address_align_bytes = 32, noc_dram_address_align_bytes = 32}], [0], [<pcie|host_mmio>], [<0, 0, 0, 0>]>} {
  func.func @forward(%arg0: tensor<64x128xf32, #layout>, %arg1: tensor<64x128xf32, #layout>) -> tensor<64x128xf32, #layout1> {
    %0 = "ttnn.open_device"() : () -> !tt.device<<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>>
    %1 = "ttnn.full"(%0) <{fillValue = 0.000000e+00 : f32}> : (!tt.device<<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>>) -> tensor<64x128xf32, #layout2>
    %2 = "ttnn.to_memory_config"(%arg0, %1) : (tensor<64x128xf32, #layout>, tensor<64x128xf32, #layout2>) -> tensor<64x128xf32, #layout2>
    %3 = "ttnn.full"(%0) <{fillValue = 0.000000e+00 : f32}> : (!tt.device<<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>>) -> tensor<64x128xf32, #layout2>
    %4 = "ttnn.to_memory_config"(%arg1, %3) : (tensor<64x128xf32, #layout>, tensor<64x128xf32, #layout2>) -> tensor<64x128xf32, #layout2>
    %5 = "ttnn.full"(%0) <{fillValue = 0.000000e+00 : f32}> : (!tt.device<<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>>) -> tensor<64x128xf32, #layout2>
    
    %6= "ttnn.generic"(%2, %4, %5) <{
      circular_buffer_attributes = [
        #tt.circular_buffer_attributes<c_in0, <0x0, 6x6>, 4096, 2048, bf16>, 
        #tt.circular_buffer_attributes<c_in1, <0x0, 6x6>, 4096, 2048, bf16>, 
        #tt.circular_buffer_attributes<c_out0, <0x0, 6x6>, 4096, 2048, bf16>
      ], 
      compute_attributes = [
        #tt.compute_attributes<
          <0x0, 6x6>, 
          "ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/compute/eltwise_binary_kernel.cpp", 
          <hifi4, false, false, false, <1>, 
          {ELTWISE_OP = "add_tiles", ELTWISE_OP_TYPE = "EltwiseBinaryType::ELWADD"}>>
      ], 
      data_movement_attributes = [
        #tt.data_movement_attributes<
          <0x0, 6x6>, 
          "ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/dataflow/reader_binary_interleaved_start_id.cpp", 
          <reader, <1, 1>>>, 
        #tt.data_movement_attributes<
          <0x0, 6x6>, 
          "ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/dataflow/writer_unary_interleaved_start_id.cpp", 
          <writer, <1, 1>>>
      ], 
      operandSegmentSizes = array<i32: 2, 1>
    }> : (tensor<64x128xf32, #layout2>, tensor<64x128xf32, #layout2>, tensor<64x128xf32, #layout2) -> tensor<64x128xf32, #layout2>
    
    %7 = "ttnn.full"(%0) <{fillValue = 0.000000e+00 : f32}> : (!tt.device<<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>>) -> tensor<64x128xf32, #layout1>
    %8 = "ttnn.to_memory_config"(%6, %7) : (tensor<64x128xf32, #layout2>, tensor<64x128xf32, #layout1>) -> tensor<64x128xf32, #layout1>
    "ttnn.close_device"(%0) : (!tt.device<<#tt.grid<8x8, (d0, d1) -> (0, d0, d1)>, [0]>>) -> ()
    return %8 : tensor<64x128xf32, #layout1>
  }
}
```

After this I went on to write a simple Python test which was supposed to mock a frontent (`tt-forge` for example) user of our compiler, who wrote their own kernels and settings for them and wanted to run it.

I managed to create pybindings for devised mlir compute, data movement and circular buffer attributes and filled them with dummy data. Printing them gives

```
[Attribute(#tt.circular_buffer_attributes<c_in0, <0x0, 6x6>, 4096, 2048, bf16>), Attribute(#tt.circular_buffer_attributes<c_in1, <0x0, 6x6>, 4096, 2048, bf16>), Attribute(#tt.circular_buffer_attributes<c_out0, <0x0, 6x6>, 4096, 2048, bf16>)]
[Attribute(#tt.data_movement_attributes<<0x0, 6x6>, "ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/dataflow/reader_binary_interleaved_start_id.cpp", <reader, <1, 1>>>), Attribute(#tt.data_movement_attributes<<0x0, 6x6>, "ttnn/cpp/ttnn/operations/eltwise/unary/device/kernels/dataflow/writer_unary_interleaved_start_id.cpp", <writer, <1, 1>>>)]
[Attribute(#tt.compute_attributes<<0x0, 6x6>, "ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/compute/eltwise_binary_kernel.cpp", <hifi4, false, false, false, <1, 1>, {ELTWISE_OP = "add_tiles", ELTWISE_OP_TYPE = "EltwiseBinaryType::ELWADD"}>>)]
```

which exactly resembles the mlir form above. Thus, user can use those attributes as objects in Python, fill them with data and crate a `.mlir` file from them.

However, I got stuck at creating `ExternalOp`. The automatically generated API was too confusing and cumbersome, I could not wrap my head around it for two days. I managed to get

```
%0 = "ttir.external_generic"(<<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>) <{operandSegmentSizes = array<i32: 2, 1>}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
```

which is close, but is missing something obviously. I left a TODO in `test/python/generic_op_test.py`. Someone should take a look at it.

**Comment by skhatriTT** [comment]
@rjakovljevicTT @kmitrovicTT @nsmithtt Is there any remaining work needed to close this issue? I'd like to understand if there are any outstanding tasks or if this can be considered complete.

**Comment by nsmithtt** [comment]
@vtangTT has recently revitalized this effort.  @vtangTT are we tracking your work with a separate issue?  If so can you please post here and then we can close this issue.

**Comment by kmitrovicTT** [comment]
This effort was paused back in august. I don't have any additional info after that.

**Comment by nsmithtt** [comment]
@vtangTT landed support in tt-metal.  Let's close this issue since it's outdated and the parts that we currently need have landed.
